File: learning_ros2/ros1_to_ros2_migration/src/ros1_ws/ros1_app_py/scripts/oop_number_counter.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import Int64
from my_robot_msgs.srv import ResetCounter


class NumberCounter:

    def __init__(self):
        self.counter_ = 0
        self.number_subscriber_ = rospy.Subscriber(
            "number", Int64, self.callback_number)
        self.reset_service_ = rospy.Service(
            "reset_counter", ResetCounter, self.callback_reset_counter)
        rospy.loginfo("Number counter has been started.")

    def callback_number(self, msg):
        self.counter_ += msg.data
        rospy.loginfo("Counter: " + str(self.counter_))

    def callback_reset_counter(self, req):
        if req.reset_value >= 0:
            self.counter_ = req.reset_value
            rospy.loginfo("Counter: " + str(self.counter_))
            return True
        return False


if __name__ == '__main__':
    rospy.init_node('number_counter')
    NumberCounter()
    rospy.spin()

File: ROS2/ros1_to_ros2_migration/src/ros1_ws/ros1_app_cpp/src/oop_number_counter.cpp
#include <ros/ros.h>
#include <std_msgs/Int64.h>
#include <my_robot_msgs/ResetCounter.h>

class NumberCounter
{

public:
    NumberCounter(ros::NodeHandle *nh)
    {
        counter_ = 0;

        number_subscriber_ = nh->subscribe("number", 1000,
                                           &NumberCounter::callback_number, this);

        reset_service_ = nh->advertiseService("reset_counter",
                                              &NumberCounter::callback_reset_counter, this);

        ROS_INFO("Number counter has been started");
    }

    void callback_number(const std_msgs::Int64 &msg)
    {
        counter_ += msg.data;
        ROS_INFO("Counter: %d", counter_);
    }

    bool callback_reset_counter(my_robot_msgs::ResetCounter::Request &req,
                                my_robot_msgs::ResetCounter::Response &res)
    {
        if (req.reset_value >= 0)
        {
            counter_ = req.reset_value;
            res.success = true;
            ROS_INFO("Counter: %d", counter_);
        }
        else
        {
            res.success = false;
        }

        return true;
    }

private:
    int counter_;
    ros::Subscriber number_subscriber_;
    ros::ServiceServer reset_service_;
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "number_counter");
    ros::NodeHandle nh;
    NumberCounter nc = NumberCounter(&nh);
    ros::spin();
    return 0;
}
